File: frontend/src/tools/forge-tools.ts
import type { AgentTool, AgentToolResult } from "@earendil-works/pi-agent-core";
import { Type, type Static, type TSchema } from "typebox";
import { getHostApi, promptAgentNamespace, type PromptAgentHostApi } from "../bridge";

const targetSchema = Type.Optional(Type.Union([
  Type.Literal("active"),
  Type.Literal("txt2img"),
  Type.Literal("img2img"),
]));

const patchSchema = Type.Object({
  operation: Type.Optional(Type.Union([
    Type.Literal("append"),
    Type.Literal("prepend"),
    Type.Literal("replace"),
    Type.Literal("replace_all"),
    Type.Literal("replace_n"),
    Type.Literal("delete"),
    Type.Literal("insert_after"),
    Type.Literal("insert_before"),
  ])),
  find: Type.Optional(Type.String({ maxLength: 20_000 })),
  replace: Type.Optional(Type.String({ maxLength: 20_000 })),
  text: Type.Optional(Type.String({ maxLength: 20_000 })),
  replacement: Type.Optional(Type.String({ maxLength: 20_000 })),
  separator: Type.Optional(Type.String({ maxLength: 32 })),
  count: Type.Optional(Type.Integer({ minimum: 1, maximum: 10_000 })),
  allow_multiple: Type.Optional(Type.Boolean()),
}, { additionalProperties: false });

const promptEditSchema = Type.Object({
  target: targetSchema,
  base_hash: Type.String({ minLength: 1, maxLength: 128 }),
  prompt: Type.Optional(Type.String({ maxLength: 50_000 })),
  diff: Type.Optional(Type.String({ maxLength: 50_000 })),
  patches: Type.Optional(Type.Array(patchSchema, { maxItems: 32 })),
  return_prompt: Type.Optional(Type.Boolean()),
}, { additionalProperties: false });

const resourceKindSchema = Type.Union([
  Type.Literal("wildcard"),
  Type.Literal("style"),
  Type.Literal("lora"),
]);

const resourceListSchema = Type.Object({
  kind: resourceKindSchema,
  query: Type.Optional(Type.String({ maxLength: 512 })),
  limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 50 })),
  cursor: Type.Optional(Type.String({ maxLength: 256 })),
}, { additionalProperties: false });

const resourceMetadataSchema = Type.Object({
  kind: resourceKindSchema,
  id: Type.String({ minLength: 1, maxLength: 512 }),
  query: Type.Optional(Type.String({ maxLength: 512 })),
  limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 100 })),
  cursor: Type.Optional(Type.String({ maxLength: 256 })),
}, { additionalProperties: false });

const danbooruSearchSchema = Type.Object({
  query: Type.Optional(Type.String({ minLength: 1, maxLength: 160 })),
  queries: Type.Optional(Type.Array(Type.String({ minLength: 1, maxLength: 160 }), { minItems: 1, maxItems: 12 })),
  category: Type.Optional(Type.String({ maxLength: 32 })),
  limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 30 })),
}, { additionalProperties: false });

const danbooruInspectSchema = Type.Object({
  name: Type.String({ minLength: 1, maxLength: 160 }),
  include_wiki: Type.Optional(Type.Boolean()),
}, { additionalProperties: false });

const danbooruInspectBatchSchema = Type.Object({
  names: Type.Array(Type.String({ minLength: 1, maxLength: 160 }), { minItems: 1, maxItems: 12 }),
  include_wiki: Type.Optional(Type.Boolean()),
}, { additionalProperties: false });

const danbooruRelatedSchema = Type.Object({
  name: Type.String({ minLength: 1, maxLength: 160 }),
  category: Type.Optional(Type.String({ maxLength: 32 })),
  limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 30 })),
}, { additionalProperties: false });

const generationParameterSchema = Type.Object({
  steps: Type.Optional(Type.Integer({ minimum: 1, maximum: 150 })),
  sampler_name: Type.Optional(Type.String({ minLength: 1, maxLength: 200 })),
  scheduler: Type.Optional(Type.String({ minLength: 1, maxLength: 200 })),
  cfg_scale: Type.Optional(Type.Number({ minimum: 0, maximum: 50 })),
  seed: Type.Optional(Type.Integer({ minimum: -1, maximum: 2_147_483_647 })),
  width: Type.Optional(Type.Integer({ minimum: 64, maximum: 4096 })),
  height: Type.Optional(Type.Integer({ minimum: 64, maximum: 4096 })),
  denoising_strength: Type.Optional(Type.Number({ minimum: 0, maximum: 1 })),
  batch_count: Type.Optional(Type.Integer({ minimum: 1, maximum: 16 })),
  batch_size: Type.Optional(Type.Integer({ minimum: 1, maximum: 16 })),
  enable_hr: Type.Optional(Type.Boolean()),
  hr_scale: Type.Optional(Type.Number({ minimum: 1, maximum: 4 })),
  hr_upscaler: Type.Optional(Type.String({ minLength: 1, maxLength: 200 })),
}, { additionalProperties: false });

const readGenerationSchema = Type.Object({
  target: targetSchema,
}, { additionalProperties: false });

const applyGenerationSchema = Type.Object({
  target: targetSchema,
  context_hash: Type.String({ minLength: 1, maxLength: 128 }),
  parameters: generationParameterSchema,
}, { additionalProperties: false });

const catalogSchema = Type.Object({
  query: Type.Optional(Type.String({ maxLength: 512 })),
  limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 50 })),
  cursor: Type.Optional(Type.String({ maxLength: 256 })),
}, { additionalProperties: false });

export const FORGE_TOOL_SCHEMAS = {
  read_prompt: Type.Object({ target: targetSchema }, { additionalProperties: false }),
  edit_prompt: promptEditSchema,
  read_negative_prompt: Type.Object({ target: targetSchema }, { additionalProperties: false }),
  edit_negative_prompt: promptEditSchema,
  list_resources: resourceListSchema,
  read_resource_metadata: resourceMetadataSchema,
  read_generation_parameters: readGenerationSchema,
  apply_generation_parameters: applyGenerationSchema,
  list_models: catalogSchema,
  list_loras: catalogSchema,
  list_embeddings: catalogSchema,
  search_danbooru_tags: danbooruSearchSchema,
  inspect_danbooru_tag: danbooruInspectSchema,
  inspect_danbooru_tags: danbooruInspectBatchSchema,
  related_danbooru_tags: danbooruRelatedSchema,
} as const;

export type ForgeToolName = keyof typeof FORGE_TOOL_SCHEMAS;
export type ForgeToolPermission = "read" | "write";

export interface ForgeToolErrorDetails {
  ok: false;
  error: {
    code: string;
    message: string;
    retryable: boolean;
  };
  result?: unknown;
}

export class ForgeToolError extends Error {
  readonly code: string;
  readonly retryable: boolean;
  readonly details: ForgeToolErrorDetails;

  constructor(code: string, message: string, retryable = false, result?: unknown) {
    super(message);
    this.name = "ForgeToolError";
    this.code = code;
    this.retryable = retryable;
    this.details = {
      ok: false,
      error: { code, message, retryable },
      ...(result === undefined ? {} : { result }),
    };
  }
}

export interface ForgeToolFactoryOptions {
  host?: () => PromptAgentHostApi | null;
  timeoutMs?: number;
  allowWrites?: () => boolean;
}

export interface ForgeAgentTool<TSchemaValue extends TSchema = TSchema> extends AgentTool<TSchemaValue, unknown> {
  readonly permission: ForgeToolPermission;
}

const TOOL_TIMEOUTS: Record<ForgeToolName, number> = {
  read_prompt: 10_000,
  edit_prompt: 15_000,
  read_negative_prompt: 10_000,
  edit_negative_prompt: 15_000,
  list_resources: 15_000,
  read_resource_metadata: 15_000,
  read_generation_parameters: 10_000,
  apply_generation_parameters: 15_000,
  list_models: 15_000,
  list_loras: 15_000,
  list_embeddings: 15_000,
  search_danbooru_tags: 30_000,
  inspect_danbooru_tag: 20_000,
  inspect_danbooru_tags: 30_000,
  related_danbooru_tags: 20_000,
};

const WRITE_TOOLS = new Set<ForgeToolName>([
  "edit_prompt",
  "edit_negative_prompt",
  "apply_generation_parameters",
]);

const defaultHost = (): PromptAgentHostApi | null => (
  typeof window === "undefined" ? null : getHostApi(promptAgentNamespace(window))
);

function abortException(): DOMException {
  return new DOMException("The Forge tool request was cancelled.", "AbortError");
}

function resultMessage(result: unknown): string {
  if (!result || typeof result !== "object") return "Forge returned an invalid tool response.";
  const value = result as Record<string, unknown>;
  const error = value.error;
  if (typeof error === "string" && error.trim()) return error;
  if (error && typeof error === "object" && typeof (error as Record<string, unknown>).message === "string") {
    return String((error as Record<string, unknown>).message);
  }
  return "The Forge tool request failed.";
}

function isFailedResult(result: unknown): boolean {
  return Boolean(result && typeof result === "object" && (result as Record<string, unknown>).ok === false);
}

async function invokeForgeTool(
  name: ForgeToolName,
  args: Record<string, unknown>,
  permission: ForgeToolPermission,
  signal: AbortSignal | undefined,
  options: ForgeToolFactoryOptions,
): Promise<unknown> {
  if (signal?.aborted) throw abortException();
  const host = (options.host ?? defaultHost)();
  if (!host || !host.isForgeAvailable()) {
    throw new ForgeToolError("forge_unavailable", "Forge is not ready, so this tool cannot run.", true);
  }
  if (permission === "write" && options.allowWrites && !options.allowWrites()) {
    throw new ForgeToolError("permission_denied", "This Forge change requires write permission.", false);
  }

  const controller = new AbortController();
  const forwardAbort = () => controller.abort();
  signal?.addEventListener("abort", forwardAbort, { once: true });
  const timeoutMs = Math.max(1_000, Math.min(60_000, options.timeoutMs ?? TOOL_TIMEOUTS[name]));
  let timer: ReturnType<typeof setTimeout> | undefined;
  let timedOut = false;
  const operation = Promise.resolve().then(() => host.executeAssistantTool({ tool: name, arguments: args }, controller.signal));
  const timeout = new Promise<never>((_, reject) => {
    timer = setTimeout(() => {
      timedOut = true;
      controller.abort();
      reject(new ForgeToolError("forge_tool_timeout", `Forge tool "${name}" timed out after ${Math.ceil(timeoutMs / 1000)} seconds.`, true));
    }, timeoutMs);
  });
  const aborted = new Promise<never>((_, reject) => {
    if (signal?.aborted) {
      controller.abort();
      reject(abortException());
      return;
    }
    signal?.addEventListener("abort", () => {
      controller.abort();
      reject(abortException());
    }, { once: true });
  });
  try {
    const result = await Promise.race([operation, timeout, aborted]);
    if (isFailedResult(result)) throw new ForgeToolError("forge_tool_failed", resultMessage(result), false, result);
    if (result === undefined) throw new ForgeToolError("invalid_forge_response", "Forge returned no result for this tool.", true);
    return result;
  } catch (error) {
    if (timedOut || error instanceof ForgeToolError || error instanceof DOMException) throw error;
    if (signal?.aborted) throw abortException();
    throw new ForgeToolError("forge_tool_failed", error instanceof Error ? error.message : "The Forge tool failed.", true, error);
  } finally {
    if (timer) clearTimeout(timer);
    signal?.removeEventListener("abort", forwardAbort);
  }
}

function textResult(result: unknown): AgentToolResult<unknown> {
  return {
    content: [{ type: "text", text: JSON.stringify(result) }],
    details: result,
  };
}

function createForgeTool<T extends TSchema>(
  name: ForgeToolName,
  label: string,
  description: string,
  parameters: T,
  permission: ForgeToolPermission,
  options: ForgeToolFactoryOptions,
  prepareArguments?: (args: Static<T>) => Record<string, unknown>,
): ForgeAgentTool<T> {
  return {
    name,
    label,
    description,
    parameters,
    permission,
    executionMode: "sequential",
    execute: async (_toolCallId, params, signal) => {
      const args = prepareArguments ? prepareArguments(params) : params as Record<string, unknown>;
      const result = await invokeForgeTool(name, args, permission, signal, options);
      return textResult(result);
    },
  };
}

export function createForgeAgentTools(options: ForgeToolFactoryOptions = {}): ForgeAgentTool[] {
  return [
    createForgeTool("read_prompt", "Read prompt", "Read the current Forge positive prompt and its latest hashes.", FORGE_TOOL_SCHEMAS.read_prompt, "read", options),
    createForgeTool(
      "edit_prompt",
      "Edit prompt",
      "Edit the positive prompt after read_prompt. Use patches/diff when the field is non-empty; full prompt overwrite is allowed only when the current field is empty.",
      FORGE_TOOL_SCHEMAS.edit_prompt,
      "write",
      options,
      (args) => ({ ...args, field: "positive" }),
    ),
    createForgeTool("read_negative_prompt", "Read negative prompt", "Read the current Forge negative prompt and its latest hash.", FORGE_TOOL_SCHEMAS.read_negative_prompt, "read", options),
    createForgeTool(
      "edit_negative_prompt",
      "Edit negative prompt",
      "Edit the negative prompt after read. Use patches/diff when non-empty; full prompt overwrite only when empty.",
      FORGE_TOOL_SCHEMAS.edit_negative_prompt,
      "write",
      options,
      (args) => ({ ...args, field: "negative" }),
    ),
    createForgeTool("list_resources", "List Forge resources", "List logical Forge resource IDs such as styles, wildcards, or LoRAs.", FORGE_TOOL_SCHEMAS.list_resources, "read", options),
    createForgeTool("read_resource_metadata", "Read resource metadata", "Read bounded metadata for one logical Forge resource.", FORGE_TOOL_SCHEMAS.read_resource_metadata, "read", options),
    createForgeTool("read_generation_parameters", "Read generation parameters", "Read the visible allowlisted Forge generation controls and context hash.", FORGE_TOOL_SCHEMAS.read_generation_parameters, "read", options),
    createForgeTool("apply_generation_parameters", "Apply generation parameters", "Apply visible allowlisted Forge generation controls with a fresh context hash.", FORGE_TOOL_SCHEMAS.apply_generation_parameters, "write", options),
    createForgeTool("list_models", "List models", "List logical Forge checkpoint IDs and labels without filesystem paths.", FORGE_TOOL_SCHEMAS.list_models, "read", options),
    createForgeTool("list_loras", "List LoRAs", "List logical LoRA IDs and labels without filesystem paths.", FORGE_TOOL_SCHEMAS.list_loras, "read", options),
    createForgeTool("list_embeddings", "List embeddings", "List logical textual inversion embedding IDs and labels without filesystem paths.", FORGE_TOOL_SCHEMAS.list_embeddings, "read", options),
    createForgeTool("search_danbooru_tags", "Search Danbooru tags", "Search live Danbooru tag candidates for one or more short visual concepts.", FORGE_TOOL_SCHEMAS.search_danbooru_tags, "read", options),
    createForgeTool("inspect_danbooru_tag", "Inspect Danbooru tag", "Inspect one Danbooru tag and optional wiki summary.", FORGE_TOOL_SCHEMAS.inspect_danbooru_tag, "read", options),
    createForgeTool("inspect_danbooru_tags", "Inspect Danbooru tags", "Inspect up to 12 selected Danbooru tags in parallel.", FORGE_TOOL_SCHEMAS.inspect_danbooru_tags, "read", options),
    createForgeTool("related_danbooru_tags", "Related Danbooru tags", "Expand one verified Danbooru seed with related tag candidates.", FORGE_TOOL_SCHEMAS.related_danbooru_tags, "read", options),
  ];
}

export function forgeToolPermission(name: string): ForgeToolPermission | null {
  return name in FORGE_TOOL_SCHEMAS ? (WRITE_TOOLS.has(name as ForgeToolName) ? "write" : "read") : null;
}

File: backend/prompt_agent/forge_tools.py
from __future__ import annotations

import re
from typing import Any


FORGE_TOOL_NAMES = (
    "read_prompt",
    "edit_prompt",
    "read_negative_prompt",
    "edit_negative_prompt",
    "list_resources",
    "read_resource_metadata",
    "read_generation_parameters",
    "apply_generation_parameters",
    "list_models",
    "list_loras",
    "list_embeddings",
    "search_danbooru_tags",
    "inspect_danbooru_tag",
    "inspect_danbooru_tags",
    "related_danbooru_tags",
)
_FORBIDDEN_KEYS = frozenset({
    "api_key",
    "apikey",
    "endpoint",
    "headers",
    "model",
    "model_id",
    "model_path",
    "modelpath",
    "mmproj_path",
    "mmprojpath",
    "llama_server_path",
    "llamaserverpath",
})
_LOCAL_PATH_RE = re.compile(r"(?:^[A-Za-z]:[\\/]|^\\\\|^/|^\.?\.?[\\/]|\.gguf$|\.safetensors$)", re.IGNORECASE)
_PATCH_OPERATIONS = frozenset({
    "append",
    "prepend",
    "replace",
    "replace_all",
    "replace_n",
    "delete",
    "insert_after",
    "insert_before",
})
_PATCH_KEYS = frozenset({
    "operation",
    "find",
    "replace",
    "text",
    "replacement",
    "separator",
    "count",
    "allow_multiple",
})


class ForgeToolValidationError(ValueError):
    """A browser-owned Forge tool request failed the server boundary."""


def validate_forge_tool_request(tool: str, payload: Any) -> dict[str, Any]:
    """Validate server-owned tool arguments without accepting provider controls."""
    if tool not in FORGE_TOOL_NAMES:
        raise ForgeToolValidationError("unknown Forge tool")
    if not isinstance(payload, dict):
        raise ForgeToolValidationError("Forge tool arguments must be an object")
    _reject_server_owned_fields(payload)
    if tool == "list_resources":
        _allow_keys(payload, {"kind", "query", "limit", "cursor"})
        _validate_resource_arguments(payload, require_id=False)
    elif tool == "read_resource_metadata":
        _allow_keys(payload, {"kind", "id", "query", "limit", "cursor"})
        _validate_resource_arguments(payload, require_id=True)
    elif tool in {"list_models", "list_loras", "list_embeddings"}:
        _allow_keys(payload, {"query", "limit", "cursor"})
        _safe_text(payload.get("query"), "query", 512)
        _bounded_integer(payload.get("limit", 20), "limit", 1, 50)
        _safe_cursor(payload.get("cursor", ""))
    elif tool in {"read_prompt", "read_negative_prompt", "read_generation_parameters"}:
        _allow_keys(payload, {"target"})
        _validate_target(payload)
    elif tool in {"edit_prompt", "edit_negative_prompt"}:
        _allow_keys(payload, {"target", "base_hash", "prompt", "diff", "patches", "return_prompt", "field"})
        _validate_target(payload)
        expected_field = "negative" if tool == "edit_negative_prompt" else "positive"
        if "field" in payload and payload["field"] != expected_field:
            raise ForgeToolValidationError(f"field must be {expected_field} for {tool}")
        _safe_text(payload.get("base_hash"), "base_hash", 128, required=True)
        _safe_text(payload.get("prompt"), "prompt", 50_000)
        _safe_text(payload.get("diff"), "diff", 50_000)
        patches = payload.get("patches")
        if patches is not None and (not isinstance(patches, list) or len(patches) > 32):
            raise ForgeToolValidationError("patches must contain at most 32 objects")
        if isinstance(patches, list):
            for index, patch in enumerate(patches):
                _validate_prompt_patch(patch, index)
        has_prompt = "prompt" in payload and payload.get("prompt") is not None
        has_diff = payload.get("diff") not in (None, "")
        has_patches = isinstance(patches, list) and len(patches) > 0
        if not has_prompt and not has_diff and not has_patches:
            raise ForgeToolValidationError("edit tools require patches, diff, or prompt")
        if has_prompt and (has_diff or has_patches):
            raise ForgeToolValidationError("prompt full overwrite cannot be combined with patches or diff")
    elif tool == "apply_generation_parameters":
        _allow_keys(payload, {"target", "context_hash", "parameters"})
        _validate_target(payload)
        _safe_text(payload.get("context_hash"), "context_hash", 128, required=True)
        parameters = payload.get("parameters")
        if not isinstance(parameters, dict):
            raise ForgeToolValidationError("parameters must be an object")
        _allow_keys(parameters, {
            "steps", "sampler_name", "scheduler", "cfg_scale", "seed", "width", "height",
            "denoising_strength", "batch_count", "batch_size", "enable_hr", "hr_scale", "hr_upscaler",
        })
        _validate_generation_parameters(parameters)
    elif tool == "search_danbooru_tags":
        _allow_keys(payload, {"query", "queries", "category", "limit"})
        _validate_danbooru_search(payload)
    elif tool == "inspect_danbooru_tag":
        _allow_keys(payload, {"name", "include_wiki"})
        _safe_text(payload.get("name"), "name", 160, required=True)
        if "include_wiki" in payload and not isinstance(payload.get("include_wiki"), bool):
            raise ForgeToolValidationError("include_wiki must be a boolean")
    elif tool == "inspect_danbooru_tags":
        _allow_keys(payload, {"names", "include_wiki"})
        names = payload.get("names")
        if not isinstance(names, list) or not names or len(names) > 12:
            raise ForgeToolValidationError("names must be a list with 1 to 12 items")
        for index, name in enumerate(names):
            _safe_text(name, f"names[{index}]", 160, required=True)
        if "include_wiki" in payload and not isinstance(payload.get("include_wiki"), bool):
            raise ForgeToolValidationError("include_wiki must be a boolean")
    elif tool == "related_danbooru_tags":
        _allow_keys(payload, {"name", "category", "limit"})
        _safe_text(payload.get("name"), "name", 160, required=True)
        _safe_text(payload.get("category"), "category", 32)
        _bounded_integer(payload.get("limit", 12), "limit", 1, 30)
    return dict(payload)


def execute_catalog_tool(tool: str, payload: dict[str, Any]) -> dict[str, Any]:
    """Return logical catalog entries without exposing server-owned paths."""
    if tool not in {"list_models", "list_loras", "list_embeddings"}:
        raise ForgeToolValidationError("unsupported Forge catalog tool")
    query = _safe_text(payload.get("query"), "query", 512)
    limit = _bounded_integer(payload.get("limit", 20), "limit", 1, 50)
    cursor = _safe_cursor(payload.get("cursor", ""))
    if tool == "list_models":
        raw_items = _model_catalog_items()
    elif tool == "list_loras":
        raw_items = _lora_catalog_items()
    else:
        raw_items = _embedding_catalog_items()
    items = []
    for item in raw_items:
        if not isinstance(item, dict):
            continue
        identifier = _logical_text(item.get("id"))
        label = _logical_text(item.get("label"))
        if identifier:
            items.append({"id": identifier, "label": label or identifier})
    if query:
        folded = query.casefold()
        items = [item for item in items if folded in f"{item['id']} {item['label']}".casefold()]
    page = items[cursor : cursor + limit]
    next_cursor = str(cursor + len(page)) if cursor + len(page) < len(items) else ""
    return {
        "ok": True,
        "kind": tool.removeprefix("list_"),
        "items": page,
        "total": len(items),
        "limit": limit,
        "cursor": str(cursor),
        "next_cursor": next_cursor,
    }


def _validate_target(payload: dict[str, Any]) -> None:
    target = payload.get("target", "active")
    if target not in {"active", "txt2img", "img2img"}:
        raise ForgeToolValidationError("target must be active, txt2img, or img2img")


def _validate_resource_arguments(payload: dict[str, Any], *, require_id: bool) -> None:
    kind = payload.get("kind")
    if kind not in {"wildcard", "style", "lora"}:
        raise ForgeToolValidationError("resource kind is invalid")
    if require_id:
        identifier = _safe_text(payload.get("id"), "id", 512, required=True)
        if _LOCAL_PATH_RE.search(identifier) or ".." in identifier.replace("\\", "/").split("/"):
            raise ForgeToolValidationError("resource id must be logical and must not be a local path")
    _safe_text(payload.get("query"), "query", 512)
    _bounded_integer(payload.get("limit", 20), "limit", 1, 100 if require_id else 50)
    _safe_cursor(payload.get("cursor", ""))


def _validate_prompt_patch(value: Any, index: int) -> None:
    if not isinstance(value, dict):
        raise ForgeToolValidationError(f"patches[{index}] must be an object")
    if set(value) - _PATCH_KEYS:
        raise ForgeToolValidationError(f"patches[{index}] contains unsupported fields")
    operation = value.get("operation", "replace")
    if operation not in _PATCH_OPERATIONS:
        raise ForgeToolValidationError(f"patches[{index}].operation is invalid")
    for field in ("find", "replace", "text", "replacement"):
        _safe_text(value.get(field), f"patches[{index}].{field}", 20_000)
    _safe_text(value.get("separator"), f"patches[{index}].separator", 32)
    if "count" in value:
        _bounded_integer(value["count"], f"patches[{index}].count", 1, 10_000)
    if "allow_multiple" in value and not isinstance(value["allow_multiple"], bool):
        raise ForgeToolValidationError(f"patches[{index}].allow_multiple must be a boolean")


def _validate_generation_parameters(value: dict[str, Any]) -> None:
    integer_ranges = {
        "steps": (1, 150),
        "seed": (-1, 2_147_483_647),
        "width": (64, 4_096),
        "height": (64, 4_096),
        "batch_count": (1, 16),
        "batch_size": (1, 16),
    }
    number_ranges = {
        "cfg_scale": (0.0, 50.0),
        "denoising_strength": (0.0, 1.0),
        "hr_scale": (1.0, 4.0),
    }
    for field, (minimum, maximum) in integer_ranges.items():
        if field in value:
            _bounded_integer(value[field], field, minimum, maximum)
    for field, (minimum, maximum) in number_ranges.items():
        if field in value:
            _bounded_number(value[field], field, minimum, maximum)
    for field in ("sampler_name", "scheduler", "hr_upscaler"):
        if field in value:
            _safe_text(value[field], field, 200, required=True)
    if "enable_hr" in value and not isinstance(value["enable_hr"], bool):
        raise ForgeToolValidationError("enable_hr must be a boolean")


def _model_catalog_items() -> list[dict[str, Any]]:
    try:
        from modules import sd_models

        result = []
        for checkpoint in getattr(sd_models, "checkpoints_list", {}).values():
            identifier = _logical_text(getattr(checkpoint, "title", "") or getattr(checkpoint, "name", ""))
            label = _logical_text(getattr(checkpoint, "short_title", "") or getattr(checkpoint, "name", ""))
            if identifier:
                result.append({"id": identifier, "label": label or identifier})
        return sorted(result, key=lambda item: item["label"].casefold())
    except (ImportError, AttributeError):
        return []


def _lora_catalog_items() -> list[dict[str, str]]:
    try:
        import networks

        result = []
        for key, entry in getattr(networks, "available_networks", {}).items():
            identifier = _logical_text(key)
            label = _logical_text(entry.get_alias() if hasattr(entry, "get_alias") else getattr(entry, "alias", key))
            if identifier:
                result.append({"id": identifier, "label": label or identifier})
        return sorted(result, key=lambda item: item["label"].casefold())
    except (ImportError, AttributeError):
        return []


def _embedding_catalog_items() -> list[dict[str, str]]:
    try:
        from modules import ui_extra_networks_textual_inversion

        database = getattr(ui_extra_networks_textual_inversion, "embedding_db", None)
        names = getattr(database, "word_embeddings", {})
        result = []
        for name in sorted(names, key=str.casefold):
            identifier = _logical_text(name)
            if identifier:
                result.append({"id": identifier, "label": identifier})
        return result
    except (ImportError, AttributeError):
        return []


def _logical_text(value: Any) -> str:
    text = str(value or "").strip().replace("\\", "/")
    if not text or _LOCAL_PATH_RE.search(text) or _local_path_fragment(text) or ".." in text.split("/"):
        return ""
    return text[:512]


def _bounded_integer(value: Any, field: str, minimum: int, maximum: int) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or not minimum <= value <= maximum:
        raise ForgeToolValidationError(f"{field} must be an integer between {minimum} and {maximum}")
    return value


def _bounded_number(value: Any, field: str, minimum: float, maximum: float) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not minimum <= value <= maximum:
        raise ForgeToolValidationError(f"{field} must be a number between {minimum:g} and {maximum:g}")
    return float(value)


def _safe_cursor(value: Any) -> int:
    if value in (None, ""):
        return 0
    if not isinstance(value, str) or not value.isdigit() or len(value) > 8:
        raise ForgeToolValidationError("cursor is invalid")
    return max(0, int(value))


def _validate_danbooru_search(payload: dict[str, Any]) -> None:
    queries = payload.get("queries")
    if queries is not None:
        if not isinstance(queries, list) or not queries or len(queries) > 12:
            raise ForgeToolValidationError("queries must be a list with 1 to 12 items")
        for index, query in enumerate(queries):
            _safe_text(query, f"queries[{index}]", 160, required=True)
    else:
        _safe_text(payload.get("query"), "query", 160, required=True)
    _safe_text(payload.get("category"), "category", 32)
    _bounded_integer(payload.get("limit", 12), "limit", 1, 30)


def _safe_text(value: Any, field: str, maximum: int, *, required: bool = False) -> str:
    if value is None:
        text = ""
    elif isinstance(value, str):
        text = value.strip()
    else:
        raise ForgeToolValidationError(f"{field} must be a string")
    if len(text) > maximum:
        raise ForgeToolValidationError(f"{field} is too large")
    if _LOCAL_PATH_RE.search(text) or _local_path_fragment(text):
        raise ForgeToolValidationError(f"{field} contains a local model or filesystem path, which is server-owned")
    if required and not text:
        raise ForgeToolValidationError(f"{field} is required")
    return text


def _reject_server_owned_fields(value: Any) -> None:
    if isinstance(value, dict):
        for key, child in value.items():
            normalized = str(key).replace("-", "_").lower()
            if normalized in _FORBIDDEN_KEYS or normalized.endswith("_path") or normalized.endswith("path"):
                raise ForgeToolValidationError("provider credentials, models, local paths, and bridge ownership fields are server-owned")
            _reject_server_owned_fields(child)
    elif isinstance(value, list):
        for child in value:
            _reject_server_owned_fields(child)


def _allow_keys(value: dict[str, Any], allowed: set[str]) -> None:
    if set(value) - allowed:
        raise ForgeToolValidationError("Forge tool arguments contain unsupported fields")


def _local_path_fragment(value: str) -> bool:
    return bool(re.search(r"(?:[A-Za-z]:[\\/]|\\\\[^\s]+|(?:^|\s)(?:\.\.?[\\/]|/))", value, re.IGNORECASE))
